src/hotspot/cpu/x86/sharedRuntime_x86.cpp

```diff
@@ -94,18 +94,49 @@ void SharedRuntime::inline_check_hashcode_from_object_header(MacroAssembler* mas
 }
 #endif //COMPILER1
 
+#ifdef _WIN64
+const juint  float_sign_mask  = 0x7FFFFFFF;
+const juint  float_infinity   = 0x7F800000;
+const julong double_sign_mask = CONST64(0x7FFFFFFFFFFFFFFF);
+const julong double_infinity  = CONST64(0x7FF0000000000000);
+#endif
```

Comment:
Isn't all this better suited for sharedRuntime_x86_64.cpp?

Reply:
You mean the entire custom frem and drem implementation unique to x64 Windows, or just this block of constants here? These constants are used by the custom Windows implementation, so they have to be in the same file. Moving the entire definition for Windows x64 to sharedRuntime\_x86\_64.cpp seems to be duplicating code even more, and I still don't know whether this is the right solution. It may be that this custom implementation may not be needed anymore and should be removed, which is why I left it in the sharedRuntime\_x86.cpp file for the time being

Reply:
I did mean the entire thing. I don't see any point in WIN64 code being in the x86 file when there is an x86_64 file. Of course I had forgotten about the Aarch64 Windows side, so maybe as I said this is really something for a os/windows file rather then `cpu/<arch>` ?

Reply:
If it's needed on AArch64 too, this should move to something in `os/windows` as David said.

However, judging from the comment in https://github.com/openjdk/jdk/blob/c40bb7621c0e49581dac587b6900b6d281572813/src/hotspot/os/windows/sharedRuntimeRem.cpp#L34-L38 this workaround might not be needed anymore.  The minimum version required for Windows-AArch64 is VS2017 (or even VS2019?), and for Windows-X64 apparently VS2022 is used (not sure what the minimum is): https://github.com/openjdk/jdk/blob/c40bb7621c0e49581dac587b6900b6d281572813/make/conf/jib-profiles.js#L1105


So maybe this can be removed altogether?  Someone needs to check if this assumption is still true today.

Alas I don't have a Windows setup anymore.  @karianna who is the local HotSpot Windows expert these days who could have a look at this? 🙂

Reply:
I don't think this workaround was intended for ARM64 in the first place, it explicitly mentions that it was written for x64, and the ifdef mess that resulted in this bug ended up bringing it over to ARM64

Reply:
The minimum Visual Studio version required to build the Windows AArch64 port is VS 2022 via the DevKit - https://wiki.openjdk.org/display/Build/Supported+Build+Platforms

I'm now verifying whether this workaround is still needed.

Reply:
@TheShermanTanker - this workaround is still needed on Windows AArch64. I haved filed https://developercommunity.visualstudio.com/t/fmod-incorrectly-returns-NaN-on-certain-/10793176 to track a fix for the underlying fmod bug.

Reply:
I'm astounded by how this happy accident looped all the way back around to being needed for ARM64, truly one of the funniest coincidences of all time

Reply:
> > I don't think this workaround was intended for ARM64 in the first place, it explicitly mentions that it was written for x64, and the ifdef mess that resulted in this bug ended up bringing it over to ARM64
> 
> So what? I'm sure it'll be fine on AArch64. Just use it.

This comment was made before Saint Wesonga confirmed that Windows/ARM64 now needed the workaround